modules/core/src/main/java/org/apache/ignite/internal/processors/diagnostic/ReconciliationExecutionContext.java

```diff
@@ -0,0 +1,152 @@
+
+    /** Jobs that were enqueued to be executed later due to permits exhaustion. */
+    private final Map<Long/*Session ID*/, Queue<ComputeJobContinuation>> pendingJobs = new LinkedHashMap<>();
+
+    /** Id of last or current reconciliation session. */
+    private long sesId;
+
+    /**
+     * @param kernalCtx Kernal context.
+     */
+    public ReconciliationExecutionContext(GridKernalContext kernalCtx) {
+        this.kernalCtx = kernalCtx;
+    }
+
+    /**
+     * @return Id of last or current reconciliation session.
+     */
+    public synchronized long sessionId() {
+        return sesId;
+    }
+
+    /**
+     * Registers new partitions reconciliation session.
+     *
+     * @param sesId Session ID.
+     * @param parallelism Parallelism level.
+     */
+    public synchronized void registerSession(long sesId, int parallelism) {
+        this.sesId = sesId;
+
+        runningJobsCnt.put(sesId, 0);
+
+        runningJobsLimit.put(sesId, parallelism);
+
+        pendingJobs.put(sesId, new LinkedList<>());
+
+        if (runningJobsCnt.size() == MAX_SESSIONS + 1) {
+            runningJobsCnt.entrySet().iterator().remove();
+
+            runningJobsLimit.entrySet().iterator().remove();
+
+            pendingJobs.entrySet().iterator().remove();
+        }
+
+
+    }
+
+    /**
+     * Acquires permit for the job execution or holds its execution if permit is not available.
+     *
+     * @param sesId Session ID.
+     * @param jobCont   Job context.
+     * @return <code>true</code> if the permit has been granted or
+     * <code>false</code> if the job execution should be suspended.
+     */
+    public synchronized boolean acquireJobPermitOrHold(long sesId, ComputeJobContinuation jobCont) {
+        if (sesId == IGNORE_JOB_PERMITS_SESSION_ID)
+            return true;
+
+        int limit = runningJobsLimit.get(sesId);
+
+        int running = runningJobsCnt.get(sesId);
+
+        if (running < limit) {
+            runningJobsCnt.put(sesId, running + 1);
+
+            return true;
+        }
+        else {
+            jobCont.holdcc();
+
+            Queue<ComputeJobContinuation> jobsQueue = pendingJobs.get(sesId);
+
+            jobsQueue.add(jobCont);
+
+            return false;
+        }
+    }
+
+    /**
+     * Releases execution permit and triggers continuation of a pending job if there are any.
+     *
+     * @param sesId Session ID.
+     */
+    public synchronized void releaseJobPermit(long sesId) {
+        if (sesId == IGNORE_JOB_PERMITS_SESSION_ID)
+            return;
+
+        int running = runningJobsCnt.get(sesId);
+
+        Queue<ComputeJobContinuation> jobsQueue = pendingJobs.get(sesId);
+
+        ComputeJobContinuation pendingJob = jobsQueue.poll();
+
+        runningJobsCnt.put(sesId, running - 1);
+
+        if (pendingJob != null) {
+            try {
+                kernalCtx.closure().runLocal(pendingJob::callcc, GridIoPolicy.MANAGEMENT_POOL);
```

Comment:
Is it work fine with the reduce jobs? Please, discase it with me.

Reply:
Yes, it will work fine.